File: features.cpp
#include <opencv.h>
#include <iostream>

using namespace std;

/*  Log
    Apr 22nd:
        - 完成第一次编写，可优化点在两个循环的位置，具体需要查询ARM处理器的指令集
        - calcHist()与【裁剪图像】尚未进行验证运行
*/
void get_context_feature(
    // input arg
    const Mat* img, 
    const int* crop_bbox,
    // output
    float* face_hist,
    float* nonface_hsit)
    {
    // 使用opencv绘制图像直方图
    // 直方图属性：
    float img_hist[256] = {0.0}; // full img hist array
    float face_hist[256] = {0.0}; // face region hist array
    float nonface_hist[256] = {0.0}; // non-face region hist array
    int histSize = 256;
    int channles = 0;
    int ranges[2] = {0, 256};
    // void calcHist(images, nimages, channels, mask, hist, dims, histSize, ranges, uniform, accumulate)
    // dims: 直方图描述的维度，若只针对灰度图，则只是1-dim，对RGB三通道都进行计算则为3-dim
    // uniform：默认为true，是否需要改为false
    // 计算全局直方图
    calcHist(img, 1, &channels, NULL, img_hist, dims=1, &histSize, ranges);
    // 根据crop_bbox坐标裁剪图像
    Rect roi(crop_bbox[0],crop_bbox[1],crop_bbox[2],crop_bbox[3]);
    Mat face = img(roi);
    // 计算人脸区域直方图
    calcHist(face, 1, &channels, NULL, img_hist, dims=1, &histSize, ranges);
    float sum = 0.0;
    // 非人脸区域直方图 = 全图直方图 - 人脸区域直方图 || 人脸区域归一化前sum的计算
    for(int i=0;i<256;i++){
        nonface_hist[i] = img_hist[i] - face_hist[i];
        sum += face_hist[i];
    }
    // 归一化
    for(int i=0;i<256;i++){
        face_hist[i] = face_hist[i]/sum;
    }

}

File: src_py/features.py
import numpy as np
import cv2
from skimage import feature
import scipy
from matplotlib import pyplot as plt


def get_context_feature(img , crop_bbox):
    img_hist =cv2.calcHist([img], [0], None, [256], [0, 256])  # 图像直方图
    # cv2.calcHist(images, channels, mask, histSize, ranges[, hist[, accumulate]]) 
    face = img[crop_bbox[0]:crop_bbox[2],crop_bbox[1]:crop_bbox[3]] # ？进行图像裁剪
    # cv2.rectangle(img, (crop_bbox[0], crop_bbox[1]),(crop_bbox[2], crop_bbox[3]), (0, 255, 0))
    # cv2.imshow('face',img)
    # cv2.waitKey(1)
    face_hist = cv2.calcHist([face], [0], None, [256], [0, 256])  # 人脸直方图
    nonface_hist = img_hist - face_hist
    face_hist = face_hist/sum(face_hist)           # 人脸直方图归一化
    nonface_hist = nonface_hist/sum(nonface_hist)  # 非人脸直方图归一化
    return face_hist.reshape(256),nonface_hist.reshape(256)


# def get_lighting_feature(img , crop_bbox):
#     face = img[crop_bbox[0]:crop_bbox[2],crop_bbox[1]:crop_bbox[3]]
#     face_hist = cv2.calcHist([face], [0], None, [256], [0, 256])  # 人脸直方图
#     face_hist = face_hist/sum(face_hist)           # 人脸直方图归一化
#     sum_x = 0
#     cnt = 0
#     index = []
#     for j in range(256):
#         if face_hist[j][0]!=0:
#             sum_x = sum_x + j
#             index.append(j)
#             cnt = cnt + 1
#     mu = sum_x / cnt
#     sigma = np.std(index)
#     x = np.linspace(1,256, 256)
#     # y = mlab.normpdf(x, mu, sigma)
#     y = scipy.stats.norm.pdf(x,mu,sigma)
#     return mu, sigma, y.reshape(256)

def get_lbp_feature(img,numPoints=8,radius=2):
    # img = cv2.equalizeHist(img)
    binsNumber = numPoints*(numPoints-1)+3 # 直方图bin的数量
    binsRange = np.arange(0,binsNumber+1) # 生成[0, binsNumber+1]的np.array，直方图x轴分隔点分布
    lbp = feature.local_binary_pattern(img, numPoints, radius, method = 'nri_uniform') #
    (hist, bins) = np.histogram(lbp.ravel(), bins=binsRange, range=(0,binsNumber),) # ？bins无用
    hist = hist/sum(hist) # 归一化
    return lbp, hist.reshape(binsNumber)